core/src/test/java/org/elasticsearch/cluster/routing/allocation/FilterAllocationDeciderTests.java

```diff
@@ -86,57 +90,80 @@ public void testFilterInitialRecovery() {
-        assertEquals(routingTable.index("idx").shard(0).primaryShard().currentNodeId(), "node2");
-
-        // replicas should be initializing
+        // first create another copy
         assertEquals(routingTable.index("idx").shard(0).replicaShards().get(0).state(), INITIALIZING);
         assertEquals(routingTable.index("idx").shard(0).replicaShards().get(0).currentNodeId(), "node1");
+        state = service.applyStartedShards(state, routingTable.index("idx").shard(0).replicaShardsWithState(INITIALIZING));
+        routingTable = state.routingTable();
+        assertEquals(routingTable.index("idx").shard(0).replicaShards().get(0).state(), STARTED);
+        assertEquals(routingTable.index("idx").shard(0).replicaShards().get(0).currentNodeId(), "node1");
+
+        // now remove the node of the other copy and fail the current
+        DiscoveryNode node1 = state.nodes().resolveNode("node1");
+        state = service.deassociateDeadNodes(
+            ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.nodes()).remove("node1")).build(),
+            true, "test");
+        state = service.applyFailedShard(state, routingTable.index("idx").shard(0).primaryShard());
 
-        // we fail it again to check if we are initializing immediately on the other node
-        state = service.applyFailedShard(state, routingTable.index("idx").shard(0).shards().get(0));
+        // now bring back node1 and see it's assigned
+        state = service.reroute(
+            ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.nodes()).add(node1)).build(), "test");
         routingTable = state.routingTable();
-        assertEquals(routingTable.index("idx").shard(0).shards().get(0).state(), INITIALIZING);
-        assertEquals(routingTable.index("idx").shard(0).shards().get(0).currentNodeId(), "node1");
+        assertEquals(routingTable.index("idx").shard(0).primaryShard().state(), INITIALIZING);
+        assertEquals(routingTable.index("idx").shard(0).primaryShard().currentNodeId(), "node1");
 
         allocation = new RoutingAllocation(allocationDeciders, state.getRoutingNodes(), state,
             null, 0, false);
         assertEquals(filterAllocationDecider.canAllocate(routingTable.index("idx").shard(0).shards().get(0),
             state.getRoutingNodes().node("node2")
-            ,allocation), Decision.YES);
+            , allocation), Decision.YES);
         assertEquals(filterAllocationDecider.canAllocate(routingTable.index("idx").shard(0).shards().get(0),
             state.getRoutingNodes().node("node1")
-            ,allocation), Decision.YES);
+            , allocation), Decision.YES);
     }
 
     private ClusterState createInitialClusterState(AllocationService service, Settings settings) {
-        boolean shrinkIndex = randomBoolean();
+        RecoverySource.Type recoveryType = randomFrom(RecoverySource.Type.EMPTY_STORE,
+            RecoverySource.Type.LOCAL_SHARDS, RecoverySource.Type.SNAPSHOT);
```

Comment:
I opted to do it like it is now for two reasons - it's more explicit so people reading can know what's going on and it offers an external validation of what's in INITIAL_RECOVERY_TYPES.